relm-derive/src/gen/gen.rs

```diff
@@ -388,6 +388,9 @@ impl<'a> Generator<'a> {
         let ident = quote! { #widget_name };
         let (properties, visible_properties) = self.gtk_set_prop_calls(widget, ident);
         let child_properties = gen_set_child_prop_calls(widget, parent, parent_widget_type, IsGtk);
+        let set_style_classes: Vec<_> = widget.style_classes.iter().map(|style_class|
+            quote_spanned! { widget_name.span() => gtk::StyleContextExt::add_class(&#widget_name.get_style_context(), &#style_class); }
```

Comment:
Is this the right span?
What happens when the trait `WidgetExt` is not in scope?

Also, could you check that the error is at the right location when you provide something else than a string in the `style_class` attribute?

Reply:
> is this the right span?

I think so, all the code around uses that span, for instance to generate the `widget.show()` and the entire block. In the end this `set_style_classes` is embedded in another quote line 405, which also uses that same span.

> What happens when the trait `WidgetExt` is not in scope?

I think it's OK. I call [add_class(&self, class_name: &str)](https://gtk-rs.org/docs/gtk/trait.StyleContextExt.html#tymethod.add_class) "manually" on `gtk::StyleContextExt`, so we don't need StyleContextExt to be in scope. In addition, I added this attribute in the `buttons-attribute.rs` test, and there the imports really seem minimal to me, and it works fine.

> Also, could you check that the error is at the right location when you provide something else than a string in the style_class attribute?

I tried it in the in the `buttons-attribute.rs` example, and this is the compilation error if I change:

```diff
 -               #[style_class="linked"]
 +               #[style_class=linked]
```

```
error: expected `,`
  --> examples/buttons-attribute.rs:88:31
   |
88 |                 #[style_class=linked]
   |                               ^^^^^^

error[E0433]: failed to resolve: use of undeclared type `Win`
   --> examples/buttons-attribute.rs:100:5
    |
100 |     Win::run(()).expect("Win::run failed");
    |     ^^^ use of undeclared type `Win`

warning: unused imports: `ButtonExt`, `Inhibit`, `LabelExt`, `OrientableExt`, `WidgetExt`
  --> examples/buttons-attribute.rs:23:5
   |
23 |     ButtonExt,
   |     ^^^^^^^^^
24 |     Inhibit,
   |     ^^^^^^^
25 |     LabelExt,
   |     ^^^^^^^^
26 |     OrientableExt,
   |     ^^^^^^^^^^^^^
27 |     WidgetExt,
   |     ^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `gtk::Orientation::Vertical`
  --> examples/buttons-attribute.rs:29:5
   |
29 | use gtk::Orientation::Vertical;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `relm::Widget`
  --> examples/buttons-attribute.rs:30:5
   |
30 | use relm::Widget;
   |     ^^^^^^^^^^^^

warning: unused import: `self::Msg::*`
  --> examples/buttons-attribute.rs:33:5
   |
33 | use self::Msg::*;
   |     ^^^^^^^^^^^^

error: aborting due to 2 previous errors; 4 warnings emitted

For more information about this error, try `rustc --explain E0433`.
error: could not compile `relm-examples`

To learn more, run the command again with --verbose.
```

This look good to me. However I also tried this:

```
                #[style_class=12]
```

and in that case the error message is not great:

```
error: custom attribute panicked
  --> examples/buttons-attribute.rs:49:1
   |
49 | #[widget]
   | ^^^^^^^^^
   |
   = help: message: Invalid style_class specification
```

However I'm not really sure what I can do to improve this :(
This is now unrelated to this code, it's related to the parsing code, and I think to parts of it that I didn't touch.